diff -up qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextcontrol.cpp.bz#1120451 qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextcontrol.cpp
--- qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextcontrol.cpp.bz#1120451	2016-11-11 07:18:01.000000000 +0100
+++ qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextcontrol.cpp	2017-07-19 17:19:18.253590767 +0200
@@ -608,7 +608,7 @@ void QQuickTextControl::clear()
 }
 
 QQuickTextControl::QQuickTextControl(QTextDocument *doc, QObject *parent)
-    : QObject(*new QQuickTextControlPrivate, parent)
+    : QInputControl(TextEdit, *new QQuickTextControlPrivate, parent)
 {
     Q_D(QQuickTextControl);
     Q_ASSERT(doc);
@@ -980,9 +980,8 @@ void QQuickTextControlPrivate::keyPressE
 
 process:
     {
-        QString text = e->text();
-        if (!text.isEmpty() && (text.at(0).isPrint() || text.at(0) == QLatin1Char('\t'))) {
-            cursor.insertText(text);
+        if (q->isAcceptableInput(e)) {
+            cursor.insertText(e->text());
             selectionChanged();
         } else {
             e->ignore();
diff -up qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextcontrol_p.h.bz#1120451 qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextcontrol_p.h
--- qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextcontrol_p.h.bz#1120451	2016-11-11 07:18:01.000000000 +0100
+++ qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextcontrol_p.h	2017-07-19 17:19:18.253590767 +0200
@@ -59,6 +59,7 @@
 #include <QtGui/qabstracttextdocumentlayout.h>
 #include <QtGui/qtextdocumentfragment.h>
 #include <QtGui/qclipboard.h>
+#include <QtGui/private/qinputcontrol_p.h>
 #include <QtCore/qmimedata.h>
 
 QT_BEGIN_NAMESPACE
@@ -71,7 +72,7 @@ class QAbstractScrollArea;
 class QEvent;
 class QTimerEvent;
 
-class Q_AUTOTEST_EXPORT QQuickTextControl : public QObject
+class Q_AUTOTEST_EXPORT QQuickTextControl : public QInputControl
 {
     Q_OBJECT
     Q_DECLARE_PRIVATE(QQuickTextControl)
diff -up qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextinput.cpp.bz#1120451 qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextinput.cpp
--- qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextinput.cpp.bz#1120451	2016-11-11 07:18:01.000000000 +0100
+++ qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextinput.cpp	2017-07-19 17:20:55.479933008 +0200
@@ -63,6 +63,7 @@
 #endif
 
 #include <QtGui/private/qtextengine_p.h>
+#include <QtGui/private/qinputcontrol_p.h>
 
 QT_BEGIN_NAMESPACE
 
@@ -2619,6 +2620,7 @@ void QQuickTextInputPrivate::init()
         option.setUseDesignMetrics(renderType != QQuickTextInput::NativeRendering);
         m_textLayout.setTextOption(option);
     }
+    m_inputControl = new QInputControl(QInputControl::LineEdit, q);
 }
 
 void QQuickTextInputPrivate::resetInputMethod()
@@ -4452,9 +4454,8 @@ void QQuickTextInputPrivate::processKeyE
     }
 
     if (unknown && !m_readOnly) {
-        QString t = event->text();
-        if (!t.isEmpty() && t.at(0).isPrint()) {
-            insert(t);
+        if (m_inputControl->isAcceptableInput(event)) {
+            insert(event->text());
             event->accept();
             return;
         }
diff -up qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextinput_p_p.h.bz#1120451 qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextinput_p_p.h
--- qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextinput_p_p.h.bz#1120451	2016-11-11 07:18:01.000000000 +0100
+++ qtdeclarative-opensource-src-5.7.1/src/quick/items/qquicktextinput_p_p.h	2017-07-19 17:19:18.254590729 +0200
@@ -70,6 +70,7 @@
 QT_BEGIN_NAMESPACE
 
 class QQuickTextNode;
+class QInputControl;
 
 class Q_QUICK_PRIVATE_EXPORT QQuickTextInputPrivate : public QQuickImplicitSizeItemPrivate
 {
@@ -227,6 +228,7 @@ public:
     QQuickItem *cursorItem;
     QQuickTextNode *textNode;
     MaskInputData *m_maskData;
+    QInputControl *m_inputControl;
 
     QList<int> m_transactions;
     QVector<Command> m_history;
