--- qtdeclarative-opensource-src-5.9.0-beta3/tools/qmlcachegen/qmlcachegen.cpp	2017-04-21 20:34:05.000000000 +0200
+++ qtdeclarative-opensource-src-5.9.0-beta3/tools/qmlcachegen/qmlcachegen.cpp.new	2017-05-06 15:20:22.055698824 +0200
@@ -41,9 +41,11 @@
 
 QT_BEGIN_NAMESPACE
 
+#if defined(V4_BOOTSTRAP)
 namespace QV4 { namespace JIT {
 Q_QML_EXPORT QV4::EvalISelFactory *createISelForArchitecture(const QString &architecture);
 } }
+#endif
 
 QT_END_NAMESPACE
 
@@ -327,7 +329,9 @@
     QScopedPointer<QV4::EvalISelFactory> isel;
     const QString targetArchitecture = parser.value(targetArchitectureOption);
 
+#if defined(V4_BOOTSTRAP)
     isel.reset(QV4::JIT::createISelForArchitecture(targetArchitecture));
+#endif
 
     if (parser.isSet(checkIfSupportedOption)) {
         if (isel.isNull())
