diff -up qtdeclarative-opensource-src-5.2.0-rc1/src/qml/jsruntime/jsruntime.pri.fedora_i686_flags qtdeclarative-opensource-src-5.2.0-rc1/src/qml/jsruntime/jsruntime.pri
--- qtdeclarative-opensource-src-5.2.0-rc1/src/qml/jsruntime/jsruntime.pri.fedora_i686_flags	2013-11-26 19:02:16.000000000 -0600
+++ qtdeclarative-opensource-src-5.2.0-rc1/src/qml/jsruntime/jsruntime.pri	2013-12-05 21:48:03.545285131 -0600
@@ -91,10 +91,10 @@ HEADERS += \
 
 # Use SSE2 floating point math on 32 bit instead of the default
 # 387 to make test results pass on 32 and on 64 bit builds.
-linux-g++*:isEqual(QT_ARCH,i386) {
-    QMAKE_CFLAGS += -march=pentium4 -msse2 -mfpmath=sse
-    QMAKE_CXXFLAGS += -march=pentium4 -msse2 -mfpmath=sse
-}
+#linux-g++*:isEqual(QT_ARCH,i386) {
+#    QMAKE_CFLAGS += -march=pentium4 -msse2 -mfpmath=sse
+#    QMAKE_CXXFLAGS += -march=pentium4 -msse2 -mfpmath=sse
+#}
 
 valgrind {
     DEFINES += V4_USE_VALGRIND
diff -up qtdeclarative-opensource-src-5.2.0-rc1/src/qml/jsruntime/qv4global_p.h.fedora_i686_flags qtdeclarative-opensource-src-5.2.0-rc1/src/qml/jsruntime/qv4global_p.h
--- qtdeclarative-opensource-src-5.2.0-rc1/src/qml/jsruntime/qv4global_p.h.fedora_i686_flags	2013-11-26 19:02:16.000000000 -0600
+++ qtdeclarative-opensource-src-5.2.0-rc1/src/qml/jsruntime/qv4global_p.h	2013-12-05 21:47:38.745545142 -0600
@@ -70,7 +70,7 @@ inline double trunc(double d) { return d
 
 // White list architectures
 
-#if defined(Q_PROCESSOR_X86)
+#if defined(Q_PROCESSOR_X86_NOT_ON_FEDORA)
 #define V4_ENABLE_JIT
 #elif defined(Q_PROCESSOR_X86_64)
 #define V4_ENABLE_JIT
